docs/source/enterprise-guide/graph-security.mdx

```diff
@@ -0,0 +1,262 @@
+
+```graphql
+query {
+  astronaut(id: "1") {
+    name
+  }
+  second:astronaut(id: "2") {
+    name
+  }
+  third:astronaut(id: "3") {
+    name
+  }
+}
+```
+
+Without any restrictions in place, clients could effectively enumerate through all nodes in a single request like the one above while slipping past other brute force protections. Limiting query breadth or using query cost analysis will help protect GraphQL APIs from this type of abuse.
+
+Another form of batching occurs when a client sends batches of full operations in a single request, which can be helpful for performance reasons in some scenarios. When operations are batched, Apollo Server receives an array of operations and sends back an array of responses to be parsed by the client (there's a batch link directly available in Apollo Client to facilitate this):
+
+```graphql
+[
+  {
+    “operationName”: "FirstAstronaut"
+    “variables":{},
+    "query":"query FirstAstronaut {\n  astronaut(id: \"1\") {\n    name\n  }\n}\n”
+  },
+  {
+    “operationName”: "SecondAstronanut"
+    “variables":{},
+    "query":"query SecondAstronanut {\n  astronaut(id: \"2\") {\n    name\n  }\n}\n”
+  },
+  {
+    “operationName”: "ThirdAstronaut"
+    “variables":{},
+    "query":"query ThirdAstronaut {\n  astronaut(id: \"3\") {\n    name\n  }\n}\n”
+  }
+]
+```
+
+With batched operations, a key consideration to keep in mind is how they may impact rate limit calculations and query cost analysis to ensure that clients won't be able to cheat rate limits through race conditions.
+
+Finally, beyond batching of fields and operations, some forms of GraphQL-related batching can help mitigate DoS attacks and generally make your API more performant overall. Even with depth limiting in place, GraphQL queries can easily lead to exponential growth of requests to backing data sources. [Dataloaders](https://github.com/graphql/dataloader) are one way that you can help make as few requests as possible to backing data sources from resolver functions within the context of a single operation.
+
+## Authentication and authorization
+
+There is no one "right" way to handle access control with GraphQL, and consolidated data graphs are no exception to this. How authentication and authorization are handled for requests to any APIs you have in use currently will likely influence how you approach these concerns with federation as well.
+
+However, one key question to address with federation is the matter of where you will validate a token sent from a client. For example, you may use middleware to intercept and validate a token in a gateway Apollo Server, add it to the request object, and then get that validated token and add it to the Apollo Server `context`:
+
+```javascript
+const server = ApolloServer({
+  gateway,
+  subscriptions: false,
```

Comment:
This line is no longer necessary in Apollo Server 3.0!

```suggestion
```

Reply:
Good catch, thanks.